src/ptwt/_util.py

```diff
@@ -245,11 +280,225 @@ def _map_result(
+        WaveletCoeff2d,
+        WaveletCoeffNd,
+    ],
+    list[int],
+]:
+    if isinstance(axes, int):
+        axes = (axes,)
+
+    torch_dtype = _check_if_tensor(coeffs[0]).dtype
+    if not _is_dtype_supported(torch_dtype):
+        raise ValueError(f"Input dtype {torch_dtype} not supported")
+
+    if ndim <= 0:
+        raise ValueError("Number of dimensions must be positive")
+
+    if tuple(axes) != tuple(range(-ndim, 0)):
+        if len(axes) != ndim:
+            raise ValueError(f"{ndim}D transforms work with {ndim} axes.")
+        else:
+            swap_fn = partial(_swap_axes, axes=axes)
+            coeffs = _map_result(coeffs, swap_fn)
+
+    # Fold axes for the wavelets
+    ds = list(coeffs[0].shape)
+    if len(ds) < ndim:
+        raise ValueError(f"At least {ndim} input dimensions required.")
+    elif len(ds) == ndim:
+        coeffs = _map_result(coeffs, lambda x: x.unsqueeze(0))
+    elif len(ds) > ndim + 1:
+        coeffs = _map_result(coeffs, lambda t: _fold_axes(t, ndim)[0])
+
+    if add_channel_dim:
+        coeffs = _map_result(coeffs, lambda x: x.unsqueeze(1))
+
+    return coeffs, ds
+
+
+# 1d case
+@overload
+def _postprocess_coeffs(
+    coeffs: list[torch.Tensor],
+    ndim: Literal[1],
+    ds: list[int],
+    axes: int,
+) -> list[torch.Tensor]: ...
+
+
+# 2d case
+@overload
+def _postprocess_coeffs(
+    coeffs: WaveletCoeff2d,
+    ndim: Literal[2],
+    ds: list[int],
+    axes: tuple[int, int],
+) -> WaveletCoeff2d: ...
+
+
+# Nd case
+@overload
+def _postprocess_coeffs(
+    coeffs: WaveletCoeffNd,
+    ndim: int,
+    ds: list[int],
+    axes: tuple[int, ...],
+) -> WaveletCoeffNd: ...
+
+
+# list of nd tensors
+@overload
+def _postprocess_coeffs(
+    coeffs: list[torch.Tensor],
+    ndim: int,
+    ds: list[int],
+    axes: Union[tuple[int, ...], int],
+) -> list[torch.Tensor]: ...
+
+
+def _postprocess_coeffs(
+    coeffs: Union[
+        list[torch.Tensor],
+        WaveletCoeff2d,
+        WaveletCoeffNd,
+    ],
+    ndim: int,
+    ds: list[int],
+    axes: Union[tuple[int, ...], int],
+) -> Union[
+    list[torch.Tensor],
+    WaveletCoeff2d,
+    WaveletCoeffNd,
+]:
+    if isinstance(axes, int):
+        axes = (axes,)
+
+    if ndim <= 0:
+        raise ValueError("Number of dimensions must be positive")
+
+    # Fold axes for the wavelets
+    if len(ds) < ndim:
+        raise ValueError(f"At least {ndim} input dimensions required.")
+    elif len(ds) == ndim:
+        coeffs = _map_result(coeffs, lambda x: x.squeeze(0))
+    elif len(ds) > ndim + 1:
+        unfold_axes_fn = partial(_unfold_axes, ds=ds, keep_no=ndim)
+        coeffs = _map_result(coeffs, unfold_axes_fn)
+
+    if tuple(axes) != tuple(range(-ndim, 0)):
+        if len(axes) != ndim:
+            raise ValueError(f"{ndim}D transforms work with {ndim} axes.")
+        else:
+            undo_swap_fn = partial(_undo_swap_axes, axes=axes)
+            coeffs = _map_result(coeffs, undo_swap_fn)
```

Comment:
I would really advise against all of these `_map_result` calls - have one function that does the processing that can be reused, then just do list comprehensions for all successive function calls.

```
coeffs = _map_result(coeffs, lambda x: x.squeeze(0))
```

will always be less readable and understandable than

```
coeffs = _map_result(coeffs)
coeffs = [coeff.squeeze(0) for coeff in coeffs]
```

when _map_result has lots of hidden functionality

Reply:
The snippet
```
coeffs = _map_result(coeffs, lambda x: x.squeeze(0))
```
applies the function `x.squeeze(0)` to all tensors in `coeffs`. In the 1d case (where `coeffs` is of type `list[Tensor]`) this is equivalent to the list comprehension, as you said. However, `coeffs` might also be
* `(Tensor, dict[str, Tensor], ...)`
* `(Tensor, (Tensor, Tensor, Tensor), ...)`

So using `_map_result` allows to write the function once for all possible coefficient types. Would it perhaps help to rename `_map_result` or add documentation for it?

Reply:
I think the name should have something to do with tree and map. The new `_apply_to_tensor_elems` name really hides how general the concept is. I would take a page from https://jax.readthedocs.io/en/latest/_autosummary/jax.tree.map.html#jax.tree.map and also use their type hinting. The concept does not exist in torch, but I think it makes sense here, since we save on a lot of boilerplate-code. Perhaps we should include the link and explain what's going on?

Reply:
Here is an interesting intro discussing the pytree processing philosophy: https://jax.readthedocs.io/en/latest/working-with-pytrees.html .

Reply:
I also think @cthoyt has a point since the tree-map concept is not very popular.

Reply:
I agree when comparing

``` python
coeffs = _map_result(coeffs, partial(torch.squeeze, 0))
````
to 
``` python
coeffs = [coeff.squeeze(0) for coeff in coeffs]
```
The list wins, but what if it's a nested structure?